File: android/src/main/kotlin/com/valentingrigorean/arcgis_maps_flutter/mapping/symbology/RendererFactory.kt
package com.valentingrigorean.arcgis_maps_flutter.mapping.symbology

import android.app.Application
import android.content.Context
import com.arcgismaps.mapping.symbology.Renderer
import com.arcgismaps.mapping.symbology.UniqueValue
import com.arcgismaps.mapping.symbology.UniqueValueRenderer

object RendererFactory {

    fun createRenderer(context: Context,data: Map<*, *>): Renderer? {
        return when (data["type"] as String) {
            "unique-value" -> parseUniqueValueRenderer(context, data)
            else -> null
        }
    }


    private fun  parseUniqueValueRenderer(context: Context, data: Map<*, *>): UniqueValueRenderer {
        val fieldsNames = data["fieldsNames"] as List<String>
        val uniqueValuesRaw = data["uniqueValues"] as List<Map<*, *>>
        val uniqueValues = uniqueValuesRaw.map {
            val description = it["description"] as String
            val label = it["label"] as String
            val symbol = SymbolFactory.createSymbol(context, it["symbol"])
            val values = it["values"] as List<Any>
            UniqueValue(description, label, symbol, values)
        }
        val defaultLabel = data["defaultLabel"] as String
        val defaultSymbol = SymbolFactory.createSymbol(context, data["defaultSymbol"])

        return UniqueValueRenderer(fieldsNames, uniqueValues, defaultLabel, defaultSymbol)
    }
}

File: ios/Classes/Mapping/Symbology/RendererFactory.swift
//
// Created by Valentin Grigorean on 18.04.2024.
//

import Foundation
import ArcGIS

class RendererFactory {
    static func createRenderer(data: Any?) -> Renderer? {
        guard let data = data as? Dictionary<String, Any> else {
            return nil
        }

        guard let type = data["type"] as? String else {
            return nil
        }
        switch type {
        case "unique-value":
            return UniqueValueRenderer(data: data)
        default:
            return nil
        }
    }
}
